src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -359,6 +361,29 @@ public UUID getFirstSourceDatasetIdFromSnapshotId(
         .orElseThrow(() -> new DatasetNotFoundException("Source dataset for snapshot not found"));
   }
 
+  public SnapshotPreviewModel retrievePreview(UUID snapshotId, String tableName, int count) {
+    Snapshot snapshot = retrieve(snapshotId);
+
+    snapshot.getTables().stream()
+        .filter(t -> t.getName().equals(tableName))
+        .findFirst()
+        .orElseThrow(
+            () ->
+                new SnapshotPreviewException(
+                    "No snapshot table exists with the name: " + tableName));
+
+    try {
+      List<Map<String, Object>> values = bigQueryPdao.getSnapshotTable(snapshot, tableName, count);
```

Comment:
Would it make sense to cache the results?  We tend to hammer on BQ in the UI and it seems like this could reduce cost (and make it a way to share snapshots with users without incurring cost)

Reply:
BQ itself will cache results in a temp table; I wouldn't add a cache here unless it becomes an issue.

Reply:
I think it's potentially going to become an issue pretty quickly, but fair enough, we could hold off on that for now

Reply:
Also, there have been conversations about long term when we share snapshots, it would be good to not incur costs for the user

Reply:
(this maybe gets actually a copy of preview data...but that is outside the scope of this ticket)

Reply:
I take your point about the parquet files. It'll not cache those. I would also consider the caching beyond the scope of the ticket. But it'll be necessary at some point.